FEA Confusion matrix derived metrics

Pawel-Kranzberg opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Continues #17265
Fixes #5516

#### What does this implement/fix? Explain your changes.
Add `metrics.specificity_score` and `metrics.npv_score`, based on `metrics.fpr_tpr_fnr_tnr_scores` from #17265

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
added a function with confusion matrix derived metrics (fpr, tpr, tnr, fnr)

Commit:
changed the true postive sum in the function

Commit:
add print

Commit:
remove one print

Commit:
remove print statements

Commit:
add coauthors.
Co-authored-by: samskruthi padigepati <https://github.com/ddhar1>
Co-authored-by: Divya Dhar<https://github.com/samskruthireddy>

Commit:
fix doc string outputs
Co-authored-by: samskruthi padigepati <https://github.com/ddhar1>
Co-authored-by: Divya Dhar<https://github.com/samskruthireddy>

Commit:
pep8 test
Co-authored-by: Divya Dhar  <https://github.com/ddhar1>
Co-authored-by: samskruthi padigepati <https://github.com/samskruthireddy>

Commit:
trivial
Co-authored-by: samskruthi padigepati <https://github.com/samskruthireddy>

Commit:
remove imported but unused flake8

Commit:
to trigger test

haochunchang committed:
Take over PR #15522

Modify doc and add deprecation to position arg.

haochunchang committed:
Modify doc and zero-division in the weighted average.

Add test for binary classification.
(Modify some lines to pass flake8)

haochunchang committed:
Add tests for binary, multiclass and empty prediction.

haochunchang committed:
Add tpr_fpr_tnr_fnr_scores to test_common.py.

haochunchang committed:
Remove pred_sum variable

haochunchang committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into confusion-matrix-derived-metrics

haochunchang committed:
Fix linting

haochunchang committed:
Fix parameter documentation

Pawel-Kranzberg committed:
Add specificity_score as a single metric

Pawel-Kranzberg committed:
Update __init__.py

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Fix linting, add what's new entry

Pawel-Kranzberg committed:
Merge branch 'confusion-matrix-derived-metrics' of https://github.com/Pawel-Kranzberg/scikit-learn into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix linting

jnothman commented on sklearn/metrics/_classification.py:
```diff
@@ -1781,6 +1984,125 @@ def recall_score(y_true, y_pred, *, labels=None, pos_label=1, average='binary',
     return r
 
 
+@_deprecate_positional_args
```
no need to deprecate on new API

jnothman commented on sklearn/metrics/_classification.py:
```diff
@@ -1543,6 +1543,209 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
     return precision, recall, f_score, true_sum
 
 
+@_deprecate_positional_args
```
no need to deprecate on new API

jnothman reviewed (commented):
specificity_score also needs adding to test_common

Pawel-Kranzberg committed:
Added scorer, fixed tests

Pawel-Kranzberg committed:
Linting

Pawel-Kranzberg committed:
Further tests

Pawel-Kranzberg committed:
Test bug fixes

Pawel-Kranzberg committed:
Test bug fixes

Pawel-Kranzberg commented:
@jnothman - Hi, will you be in position to review this PR?

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Update v1.0.rst

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

jnothman commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1962,6 +2105,35 @@ def test_fscore_warnings(zero_division):
+                          average='micro', zero_division=zero_division)
+        if zero_division == "warn":
+            assert (str(record.pop().message) ==
+                    'Tnr is ill-defined and '
```
Tnr -> TNR

jnothman reviewed (commented):
Sorry, I'd forgotten to submit this partial review a few weeks ago

Pawel-Kranzberg committed:
Add `metrics.fpr_tpr_fnr_tnr_scores` to v1.0.rst

Pawel-Kranzberg commented on doc/whats_new/v1.0.rst:
```diff
@@ -207,6 +207,10 @@ Changelog
   for non-English characters. :pr:`18959` by :user:`Zero <Zeroto521>`
   and :user:`wstates <wstates>`.
 
+- |Enhancement| Add `metrics.specificity_score` function. 
```
Added now.

Pawel-Kranzberg committed:
tpr, fpr, tnr, fnr to upper case

Pawel-Kranzberg committed:
Update test_classification.py

Pawel-Kranzberg committed:
Linting fix

Pawel-Kranzberg commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -332,6 +334,145 @@ def test_precision_recall_f_ignored_labels():
+    y_true = np.array([[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 1, 0]])
+    y_pred = np.array([[0, 0, 0, 0], [0, 0, 0, 1], [0, 1, 1, 0]])
+
+    zero_division = 1.0 if zero_division == 1.0 else 0.0
```
I hope the new version is clearer.

Pawel-Kranzberg commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1962,6 +2105,35 @@ def test_fscore_warnings(zero_division):
+                       np.array([[1, 1], [1, 1]]),
+                       average='micro', zero_division=zero_division)
+    with warnings.catch_warnings(record=True) as record:
+        warnings.simplefilter('always')
```
Is the new version better?

Pawel-Kranzberg commented:
> Sorry, I'd forgotten to submit this partial review a few weeks ago

No worries. Please have a look at the current version.

Pawel-Kranzberg committed:
Typo fix

Pawel-Kranzberg committed:
Revert "Typo fix"

This reverts commit 5a391fa47c7fbe3328657073cd89475f86506efc.

Pawel-Kranzberg committed:
retrigger checks

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
retrigger checks

Pawel-Kranzberg committed:
retrigger checks

Pawel-Kranzberg committed:
retrigger checks

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

glemaitre commented:
@jnothman I am wondering if are not duplicating code here. I am thinking that we could have a single private function that could compute all the necessary statistics and then the precision-recall and other metrics would only call this function.
I could even think that we could pass a parameter that is a dictionary that could explicitly request some of the metrics to be returned as a bunch.

Here, I am also questioning adding the tpr/fpr/fnr/tnr function since it will have duplicated statistics with the recall score and potentially later the specificity. It seems that @amueller wanted to avoid this in the original issue: https://github.com/scikit-learn/scikit-learn/issues/5516#issuecomment-306464383

jnothman commented:
multilabel_confusion_matrix removed a lot of the duplication. Problem is that users want these metrics and for whatever reason find them hard to compute themselves. And given how much revision we've made to P/R/F, there are certainly tricky cases! The present PR could have less duplication; but the main place there is duplication here is in the tests which will require more work to refactor.

I had thought that to avoid clutter, we could avoid putting these measures into `metrics`, but make them available when a user sought the set of scorers relevant to a particular task vis a vis #12385

glemaitre commented:
@Pawel-Kranzberg Could you solve the conflicts in the files. I assume that the conflicts here are just related to execute `black` on the file.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    Parameters
+    ----------
+    y_true : {array-like, label indicator array, sparse matrix} \
+        of shape (n_samples,)
```
```suggestion
            of shape (n_samples,)
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    Returns
+    -------
+    tpr : float (if average is not None), \
+        or ndarray of shape (n_unique_labels,)
```
```suggestion
            or ndarray of shape (n_unique_labels,)
        True positive ratio.
```

You can modify the next entries in the same manner.

Pawel-Kranzberg replied:
Done.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    >>> y_true = np.array(['cat', 'dog', 'pig', 'cat', 'dog', 'pig'])
+    >>> y_pred = np.array(['cat', 'pig', 'dog', 'cat', 'cat', 'dog'])
+    >>> tpr_fpr_tnr_fnr_scores(y_true, y_pred, average='macro')
+    (0.3333333333333333, 0.3333333333333333, 0.6666666666666666,
+     0.6666666666666666)
```
```suggestion
    (0.33..., 0.33..., 0.66..., 0.66...)
```

glemaitre replied:
You can use the ellipsis for the other example as well.

Pawel-Kranzberg replied:
Done.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    (0.3333333333333333, 0.3333333333333333, 0.6666666666666666,
+    0.6666666666666666)
+
+    It is possible to compute per-label fpr, fnr, tnr, tpr and
```
```suggestion
    It is possible to compute per-label FPR, FNR, TNR, TPR and
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    -----
+    When ``true positive + false negative == 0``, TPR, FNR are undefined;
+    When ``true negative + false positive == 0``, FPR, TNR are undefined.
+    In such cases, by default the metric will be set to 0, as will f-score,
```
```suggestion
    In such cases, by default the metric will be set to 0, as will F-score,
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+        or ndarray of shape (n_unique_labels,)
+        The number of occurrences of each label in ``y_true``.
+
+    References
```
We would need a See Also section similar to other metrics.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1776,6 +1978,126 @@ def recall_score(y_true, y_pred, *, labels=None, pos_label=1, average='binary',
+                      zero_division="warn"):
+    """Compute specificity, also known as true negative rate.
+
+    The specificity is the ratio ``tn / (tn + fp)`` where ``tn`` is the number
```
You as well single backticks instead of double backticks here and after.

Pawel-Kranzberg replied:
Done.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    returns the true positive rate, false positive rate, true negative rate
+    and false negative rate if ``average`` is one of ``'micro'``, ``'macro'``,
+    ``'weighted'`` or ``'samples'``.
+
```
We will need to add some documentation in the user guide and linked it.

glemaitre replied:
You can as well add the sphinx directive:

```python
.. versionadded:: 1.1
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1776,6 +1978,126 @@ def recall_score(y_true, y_pred, *, labels=None, pos_label=1, average='binary',
+    The specificity is intuitively the ability of the classifier to find
+    all the negative samples.
+
+    The best value is 1 and the worst value is 0.
```
We will need to add some documentation in the user guide and linked it.

glemaitre replied:
We should also add the sphinx directive:

```python
.. versionadded:: 1.1
```

Pawel-Kranzberg replied:
Done.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1776,6 +1978,126 @@ def recall_score(y_true, y_pred, *, labels=None, pos_label=1, average='binary',
+
+    Parameters
+    ----------
+    y_true : 1d array-like, or label indicator array / sparse matrix
```
Could you use the same way of writing docstring as in the previous function with the `{}` characters

glemaitre replied:
and providing the shape of the array

Pawel-Kranzberg replied:
Done.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1776,6 +1978,126 @@ def recall_score(y_true, y_pred, *, labels=None, pos_label=1, average='binary',
+
+    Returns
+    -------
+    specificity : float (if average is not None) or array of float of shape
+        (n_unique_labels,)
```
```suggestion
    specificity : float or ndarray of shape (n_unique_labels,), dtype=np.float64
```
We can add the information in the description for when it will be a float and when it will be an array.

Pawel-Kranzberg replied:
Done.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1776,6 +1978,126 @@ def recall_score(y_true, y_pred, *, labels=None, pos_label=1, average='binary',
+
+    See Also
+    --------
+    classification_report, precision_recall_fscore_support, recall_score,
```
It would be best to have an entry per line and a short description for each entry

Pawel-Kranzberg replied:
Done.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -2171,7 +2493,8 @@ def log_loss(y_true, y_pred, *, eps=1e-15, normalize=True, sample_weight=None,
 
-    y_pred : array-like of float, shape = (n_samples, n_classes) or (n_samples,)
+    y_pred : array-like of float, shape = (n_samples, n_classes) \
+    or (n_samples,)
```
```suggestion
            or (n_samples,)
```

glemaitre reviewed (commented):
a first round of comment regarding the documentation.

Scoodood commented:
The current `metrics.classification_report` return
- Precision (Positive Predictive Value)
- Recall (True Positive Rate)
- F1
- Support

However, it is still incomplete. We still need
- **Negative Predictive Value (NPV)**
- **Specificity (True Negative Rate)**

The `metrics.specificity_score` is a great addition! 
Please also added `metrics.npv_score`, and include both `specificity` and `npv` into the classification report.

In the future, it would be nice to include these as well
- Prevalence
- MCC
- Kappa
- AUROC (Area under ROC curve)
- AUPRC (Area under PR curve)

to cover most benchmarking angles. Please [see here](https://en.wikipedia.org/wiki/Precision_and_recall)

glemaitre commented:
@Scoodood I would like to focus this PR only on the metrics addition without touching on the classification report. It could be done in a subsequent PR. Could you open an issue from now on mentioning your feature request regarding the report? I assume that at some point, it could be interesting to have a parameter to filter the desired metrics.

glemaitre commented:
@Pawel-Kranzberg We could consider to add the NPV as well to close the loop.

Scoodood commented:
> @Scoodood I would like to focus this PR only on the metrics addition without touching on the classification report. It could be done in a subsequent PR. Could you open an issue from now on mentioning your feature request regarding the report? I assume that at some point, it could be interesting to have a parameter to filter the desired metrics.

It's done. https://github.com/scikit-learn/scikit-learn/issues/21000

Pawel-Kranzberg committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Fix recall

Pawel-Kranzberg committed:
Linting fixes

Pawel-Kranzberg committed:
Fix linting

Fix linting

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix linting

Fix linting

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix linting

Add specifity test to test_ovr_multilabel_dataset

Fix linting

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix linting

Fix linting

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix docs formatting

Pawel-Kranzberg committed:
Fix linting in _classification.py

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Linting troubleshooting

Fix linting in _classification.py

FIx linting in _classification.py

Troubleshooting test of _classification.py linting

Fix linting

Pawel-Kranzberg commented:
> @Pawel-Kranzberg Could you solve the conflicts in the files. I assume that the conflicts here are just related to execute `black` on the file.

Thank you. I'm currently still working on black-compliance in _classification.py . 
My local black claims that the "file would be left unchanged", but the Azure one disagrees.

Pawel-Kranzberg committed:
Doc fixes in _classification.py

Pawel-Kranzberg committed:
Revert change to v1.0.rst

Pawel-Kranzberg committed:
Revert "Revert change to v1.0.rst"

This reverts commit 17ad4fa7358b937cfbce3827210bba3afc7afd97.

Pawel-Kranzberg commented on doc/whats_new/v1.0.rst:
```diff
@@ -366,6 +366,11 @@ Changelog
   quantile regression. :pr:`19415` by :user:`Xavier Dupré <sdpython>`
   and :user:`Oliver Grisel <ogrisel>`.
+  
+- |Feature| Add :func:`metrics.fpr_tpr_fnr_tnr_scores`
```
Done.

Pawel-Kranzberg replied:
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+                                                   'tnr', 'fnr'),
+                           sample_weight=None, zero_division="warn"):
+    """Compute True Positive Rate (TPR), False Positive Rate (FPR),\
+    True Negative Rate (TNR), False Negative Rate (FNR) for each class
```
Done.

Pawel-Kranzberg commented:
> @Pawel-Kranzberg Could you solve the conflicts in the files. I assume that the conflicts here are just related to execute `black` on the file.

Done.

Pawel-Kranzberg committed:
Update v1.0.rst

Pawel-Kranzberg committed:
Update v1.1.rst

Pawel-Kranzberg committed:
Update _classification.py

Update docs of tpr_fpr_tnr_fnr_scores and specificity_score

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    """Compute True Positive Rate (TPR), False Positive Rate (FPR),\
+    True Negative Rate (TNR), False Negative Rate (FNR) for each class
+
+    The TPR is the ratio ``tp / (tp + fn)`` where ``tp`` is the number of
```
Done

Pawel-Kranzberg committed:
Update _classification.py

Update docs of tpr_fpr_tnr_fnr_scores and specificity_score

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Update docs in _classification.py

Pawel-Kranzberg committed:
Add npv_score

Pawel-Kranzberg committed:
Add npv_score

Pawel-Kranzberg committed:
Add npv_score

Pawel-Kranzberg committed:
Add npv_score

Pawel-Kranzberg committed:
npv_score

Pawel-Kranzberg committed:
Add npv_score

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -1540,6 +1540,208 @@ def precision_recall_fscore_support(y_true, y_pred, *, beta=1.0, labels=None,
+    """Compute True Positive Rate (TPR), False Positive Rate (FPR),\
+    True Negative Rate (TNR), False Negative Rate (FNR) for each class
+
+    The TPR is the ratio ``tp / (tp + fn)`` where ``tp`` is the number of
```
Done.

Pawel-Kranzberg committed:
Add `npv_score`

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Add `npv_score`

Pawel-Kranzberg committed:
Fix linting

Pawel-Kranzberg committed:
Fix test_common.py

Pawel-Kranzberg committed:
Add `npv_score`

Pawel-Kranzberg commented:
> @Pawel-Kranzberg We could consider adding the NPV as well to close the loop.

Done.

Pawel-Kranzberg committed:
Update docs

Pawel-Kranzberg committed:
Update model_evaluation.rst

Pawel-Kranzberg committed:
Update _classification.py

Pawel-Kranzberg committed:
Fix docs

Pawel-Kranzberg committed:
Fix model_evaluation.rst

Pawel-Kranzberg committed:
Fix docs

Pawel-Kranzberg commented:
@glemaitre - Thank you for the comments. The PR is ready for another round of review.

Pawel-Kranzberg commented:
@Scoodood - FYI, I've added the NPV score fuction (`npv_score`).

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Scoodood commented:
> @Scoodood - FYI, I've added the NPV score fuction (`npv_score`).

hi @Pawel-Kranzberg, just wondering when these new metrics will be released for public consumption? The latest official scikit-learn still doesn't have these metrics yet.

Pawel-Kranzberg commented:
> > @Scoodood - FYI, I've added the NPV score fuction (`npv_score`).
> 
> hi @Pawel-Kranzberg, just wondering when these new metrics will be integrated into the official scikit-learn for public consumption?

I keep my fingers crossed...

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        fpr = np.average(fpr, weights=weights)
+        tnr = np.average(tnr, weights=weights)
+        fnr = np.average(fnr, weights=weights)
+    return tpr, fpr, tnr, fnr
```
This comment is orthogonal to this PR and will not require any action.
I am wondering about the API here and the compatibility with multimetric.

I think that it would be nice to support multimetric in `GridSearchCV` with the minumum of code. I would expect the following most probably:

```python
scorer = make_scorer(tpr_fpr_tnr_fnr_score)
GridSearchCV(..., scoring=scorer)
```

It would only work if `make_scorer` was able to return a dict I think. I am wondering if we could have an API that add a `return_type` to the metric function returning multiple scalars such that we can decide to return an array (2D as in confusion matrix), a tuple (like here and other metrics) or a Python dict with the name and the metric associated.

In this manner, `make_scorer` can understand `dict` and output a dict that can directly be used by the `scoring` of the `GridSearchCV` without the need of defining a custom callable to simply rewrap the scalar into a `dict`.

@jnothman do you have any thought. I am almost sure that this aleady something that you might have discussed when creating the multimetric API

glemaitre reviewed (commented):
This is a partial review. I will look at the test later today.

glemaitre commented on doc/modules/model_evaluation.rst:
```diff
@@ -806,6 +810,18 @@ precision-recall curve as follows.
 
+Precision can also be referred to as the `positive predictive value (PPV)
+<https://en.wikipedia.org/wiki/Positive_and_negative_predictive_values>`_,
+e.g. in the context of bioscience. A closely related metric is 
```
There is a trailing space at the end of the line

Pawel-Kranzberg replied:
Done.

glemaitre commented on doc/modules/model_evaluation.rst:
```diff
@@ -806,6 +810,18 @@ precision-recall curve as follows.
+Recall can also be called the hit rate, or true positive rate (TPR). Especially
+in biostatistics, it is also known as `sensitivity <https://en.wikipedia.org/wiki/Sensitivity_and_specificity>`_
+, which is related to `specificity <https://en.wikipedia.org/wiki/Sensitivity_and_specificity>`_.
+In turn, specificity is also referred to as selectivity, or true negative rate (TNR), 
```
trailing space at the end of the line

Pawel-Kranzberg replied:
Done.

glemaitre commented on doc/modules/model_evaluation.rst:
```diff
@@ -806,6 +810,18 @@ precision-recall curve as follows.
         :scale: 75
```
I am thinking that we should document the function returning the ratios.
I think that we should move the confusion matrix presentation before the accuracy score that use the TP FP TN FR already. Like this we could document the ratio function just after the confusion matrix.

Pawel-Kranzberg replied:
Attempted.

glemaitre commented on doc/modules/model_evaluation.rst:
```diff
@@ -806,6 +810,18 @@ precision-recall curve as follows.
         :scale: 75
         :align: center
 
+Precision can also be referred to as the `positive predictive value (PPV)
```
I am thinking that we could isolate these two metrics in a new section where we can provide more details regarding binary and multiclass case and the effect of the averaging similarly to the precision-recall section

Pawel-Kranzberg replied:
Attempted.

glemaitre reviewed (commented):
A couple of more reviews

Pawel-Kranzberg committed:
Update _classification.py

Implement review suggestions.

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+):
+    """Compute the negative predictive value (NPV).
+
+    .. versionadded:: 1.1
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        setting `labels=[pos_label]` and `average != "binary"` will report
+        scores for that label only.
+
+    average : str, {None, "binary", "micro", "macro", "samples", "weighted"}, \
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        scores for that label only.
+
+    average : str, {None, "binary", "micro", "macro", "samples", "weighted"}, \
+        default=None
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+):
+    """Compute the TPR, FPR, TNR, FNR for each class.
+
+    .. versionadded:: 1.1
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+
+    labels = _check_set_wise_labels(y_true, y_pred, average, labels, pos_label)
+
+    # Calculate tp_sum, fp_sum, tn_sum, fn_sum, pos_sum, neg_sum
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        setting `labels=[pos_label]` and `average != "binary"` will report
+        scores for that label only.
+
+    average : str, {None, "binary", "micro", "macro", "samples", "weighted"} \
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+
+    See Also
+    --------
+    classification_report : A text report showing the key classification metrics.
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        weights = None
+
+    if average is not None:
+        assert average != "binary" or len(fpr) == 1
```
Attempted.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+    else:
+        weights = None
+    if average is not None:
+        assert average != "binary" or len(NPV) == 1
```
Attempted.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        setting `labels=[pos_label]` and `average != "binary"` will report
+        scores for that label only.
+
+    average : str, {None, "binary", "micro", "macro", "samples", "weighted"} \
```
Done.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+    precision_recall_fscore_support : The key classification metrics.
+    precision_score : Precision or positive predictive value (PPV).
+    recall_score : Recall, sensitivity, hit rate, or true positive rate (TPR).
+    multilabel_confusion_matrix : Confusion matrices for each class or sample.
```
I'm not sure I understand, hence I've left only the precision_score and recall_score, as hinted above.

Pawel-Kranzberg commented on sklearn/metrics/_classification.py:
```diff
@@ -2649,3 +2649,547 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
     return np.average((y_true - y_prob) ** 2, weights=sample_weight)
+
+
+def tpr_fpr_tnr_fnr_scores(
```
Done.

Pawel-Kranzberg committed:
Implement review

Implement review

Pawel-Kranzberg committed:
Fix linting

Fix linting

Pawel-Kranzberg committed:
Update v1.1.rst

Fix linting

Pawel-Kranzberg committed:
Update _classification.py

Fix docs

Pawel-Kranzberg committed:
Update _classification.py

Fix docs

Pawel-Kranzberg committed:
Update _classification.py

Fix docs.

Pawel-Kranzberg commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -348,6 +351,133 @@ def test_precision_recall_f_ignored_labels():
+    with pytest.raises(ValueError):
+        tpr_fpr_tnr_fnr_scores(y_true, y_pred, average="samples")
+
+    # same prediction but with and explicit label ordering
```
The coverage is at 100%.

Pawel-Kranzberg replied:
> It makes me think that we should check the coverage. I don't see the report in the last push. It could be nice to be sure that we covered all the branch of the code.

The coverage is proudly at 100%.

Pawel-Kranzberg replied:
> Could we add a case where we take a subset of the labels.

Done.

Pawel-Kranzberg replied:
> We should probably as well test when there is too many labels given

Done.

Pawel-Kranzberg committed:
Update the tpr_fpr_tnr_fnr_score doc and specificity test

Pawel-Kranzberg committed:
Update test_classification.py

Fix test_specificity_warnings

Pawel-Kranzberg committed:
Update test_classification.py

Fix test_specificity_warnings

Pawel-Kranzberg commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -2059,6 +2190,95 @@ def test_fscore_warnings(zero_division):
+        zero_division=zero_division,
+    )
+    if zero_division == "warn":
+        pytest.warns(Warning if zero_division == "warn" else None)
```
Done.

Pawel-Kranzberg committed:
Update test_classification.py

Add tests for subsets of and too many of labels.

Pawel-Kranzberg committed:
Update test_classification.py

fix linting

Pawel-Kranzberg committed:
Update test_classification.py

Fix test_tpr_fpr_tnr_fnr_score_extra_labels()

Pawel-Kranzberg committed:
Update test_classification.py

Moar tests.

Pawel-Kranzberg committed:
Update test_classification.py

Fix linting.

Pawel-Kranzberg committed:
Update test_classification.py

Fix linting.

Pawel-Kranzberg commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -2059,6 +2190,95 @@ def test_fscore_warnings(zero_division):
+        npv_score(y_true, y_pred, average="samples")
+
+    # same prediction but with and explicit label ordering
+    assert_array_almost_equal(
```
Done.

Pawel-Kranzberg committed:
Update test_classification.py

Expand test_tpr_fpr_tnr_fnr_score_extra_labels()

Pawel-Kranzberg commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -348,6 +351,133 @@ def test_precision_recall_f_ignored_labels():
+    assert_array_almost_equal(tnr, 0.8, 2)
+    assert_array_almost_equal(fnr, 0.47, 2)
+
+    with pytest.raises(ValueError):
```
Attempted.

Pawel-Kranzberg committed:
Update test_classification.py

Improvements of tests.

Pawel-Kranzberg committed:
Update test_classification.py

Fix tests.

Pawel-Kranzberg committed:
Update test_classification.py

Fix tests.

Pawel-Kranzberg committed:
Update test_classification.py

Enhance tests.

Pawel-Kranzberg committed:
Update test_classification.py

Fix tests.

Pawel-Kranzberg committed:
Update test_classification.py

typo fix

Pawel-Kranzberg committed:
Update test_classification.py

Test enhancements.

Pawel-Kranzberg committed:
Update test_classification.py

Downgrade test_tpr_fpr_tnr_fnr_score_with_an_empty_prediction()

Pawel-Kranzberg commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -2059,6 +2190,95 @@ def test_fscore_warnings(zero_division):
+        zero_division=zero_division,
+    )
+    if zero_division == "warn":
+        pytest.warns(Warning if zero_division == "warn" else None)
```
Attempted.

Pawel-Kranzberg committed:
Update model_evaluation.rst

Expand and reorder docs.

Pawel-Kranzberg committed:
Update model_evaluation.rst

Expand docs.

Pawel-Kranzberg committed:
Update model_evaluation.rst

Update docs.

Pawel-Kranzberg committed:
Update model_evaluation.rst

Fix docs.

Pawel-Kranzberg committed:
Update model_evaluation.rst

Fix docs.

Pawel-Kranzberg committed:
Update model_evaluation.rst

Fix docs.

glemaitre commented:
Nice the CI is passing. I will make a new pass on it.

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Update model_evaluation.rst

Align the PR docs with the main branch.

Pawel-Kranzberg committed:
Update v1.2.rst

Pawel-Kranzberg committed:
Update test_classification.py

Align test_npv_warnings() and test_tpr_fpr_tnr_fnr_score_with_an_empty_prediction() with pytest 7.

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Update test_classification.py

Fix test_tpr_fpr_tnr_fnr_score_with_an_empty_prediction()

Pawel-Kranzberg committed:
Update test_classification.py

Fix test_tpr_fpr_tnr_fnr_score_with_an_empty_prediction().

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

jeremiedbb commented:
We won't have time to finish the review on this one before the 1.2 release. Moving it to 1.3

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Update test_classification.py

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Update test_multiclass.py

Linting fix

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Update v1.4.rst

Pawel-Kranzberg committed:
Update v1.2.rst

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg commented:
![image](https://github.com/scikit-learn/scikit-learn/assets/7954786/64bff819-a513-4fb4-bc7a-7210ca135123)

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

adrinjalali commented:
@lorentzenchr would you be able to maybe have a look?

lorentzenchr commented:
From the long discussion in #5516, better ping @jnothman @amueller @glemaitre .

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Fix linting in test_classification.py

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

lorentzenchr commented:
@Pawel-Kranzberg Could you please stop merging the main branch as long as there are no merge conflict? The reason is that it might hide discussions in Github and also generates notifications. And it should be in your interest that maintainers don't unsubscribe from this PR.

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Fix linting in _classification.py

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

JoshuaDBruton commented:
Hi everyone, I have need for specificity_score as well and it would be convenient if it worked like Recall and the others that are available. I have added it locally and was considering opening a PR but saw this one was open already.

Just seconding that this would be a useful thing.

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Remove the legacy assert_no_warnings in test_classification.py

Pawel-Kranzberg committed:
Convert score values to float in _classification.py

Pawel-Kranzberg committed:
Merge branch 'main' into confusion-matrix-derived-metrics

Pawel-Kranzberg committed:
Update _classification.py

Pawel-Kranzberg committed:
Update _classification.py
